src/main/java/org/opensearch/geospatial/index/mapper/xypoint/XYPointParser.java

```diff
@@ -71,101 +78,158 @@ public static XYPoint parseXYPoint(Object value, final boolean ignoreZValue) thr
-     * @throws IOException
-     */
-    private static XYPoint parseXYPointArray(XContentParser subParser, final boolean ignoreZValue, double x, double y) throws IOException {
-        XYPoint point = new XYPoint();
-        int element = 0;
-        while (subParser.nextToken() != XContentParser.Token.END_ARRAY) {
-            if (subParser.currentToken() != XContentParser.Token.VALUE_NUMBER) {
-                throw new OpenSearchParseException("numeric value expected");
+    private static XYPoint parseGeoJsonFields(final XContentParser parser, final XYPoint point, final boolean ignoreZValue)
+        throws IOException {
+        final int numberOfFields = 2;
+        boolean hasTypePoint = false;
+        boolean hasCoordinates = false;
+        for (int i = 0; i < numberOfFields; i++) {
+            if (i != 0) {
+                parser.nextToken();
             }
-            element++;
-            if (element == 1) {
-                x = subParser.doubleValue();
-            } else if (element == 2) {
-                y = subParser.doubleValue();
-            } else if (element == 3) {
-                XYPoint.assertZValue(ignoreZValue, subParser.doubleValue());
+
+            if (parser.currentToken() != XContentParser.Token.FIELD_NAME) {
+                if (hasTypePoint == false) {
+                    throw new OpenSearchParseException("field [{}] missing", GEOJSON_TYPE);
+                }
+                if (hasCoordinates == false) {
+                    throw new OpenSearchParseException("field [{}] missing", GEOJSON_COORDS);
+                }
+            }
+
+            if (GEOJSON_TYPE.equals(parser.currentName())) {
+                if (parser.nextToken() != XContentParser.Token.VALUE_STRING) {
+                    throw new OpenSearchParseException("{} must be a string", GEOJSON_TYPE);
+                }
+
+                // To be consistent with geo_shape parsing, ignore case here as well.
+                if (ShapeType.POINT.name().equalsIgnoreCase(parser.text()) == false) {
+                    throw new OpenSearchParseException("{} must be Point", GEOJSON_TYPE);
+                }
+                hasTypePoint = true;
+            } else if (GEOJSON_COORDS.equals(parser.currentName())) {
+                if (parser.nextToken() != XContentParser.Token.START_ARRAY) {
+                    throw new OpenSearchParseException("{} must be an array", GEOJSON_COORDS);
+                }
+                parseXYPointArray(parser, point, ignoreZValue);
+                hasCoordinates = true;
             } else {
-                throw new OpenSearchParseException("[xy_point] field type does not accept more than 3 dimensions");
+                throw new OpenSearchParseException(ERR_MSG_INVALID_FIELDS);
             }
         }
-        return point.reset(x, y);
+
+        return point;
+    }
+
+    private static XYPoint parseXYPointArray(final XContentParser parser, final XYPoint point, final boolean ignoreZValue)
```

Comment:
add java doc on top of this function specifying what exactly you are parsing.

Reply:
ack